Improve service instantiation & use Doctrine DBAL instead of Zend DB in CacheManager

pascalheidmann-bedarf opened the pull request:
### 1. Why is this change necessary?
Several services are instantiated with parameter "service_container" which can basically be any `Container` that implements the `ContainerInterface` - therefore lots of service classes typehint to narrow with just `Container`.
Then there are several instances where the container is passed to the constructor to load other services. These cases can be simplified a lot.

### 2. What does this change do, exactly?
- Replace all typehints of `Container` with the interface `ContainerInterface` if possible.
- Instantiate services with their required services directly instead of going the route of loading them manually via container
  - if possible remove container completely from services 
- As a bonus I updated the CacheManager to use the `dbal_connection` service instead of `db`.

### 3. Describe each step to reproduce the issue or behaviour.


### 4. Please link to the relevant issues (if any).


### 5. Which documentation changes (if any) need to be made because of this PR?
The PR changes the factory method signature of `Shopware\Components\DependencyInjection\Bridge\Plugin::factory`

### 6. Checklist

- [x] I have written tests and verified that they fail without my change
- [x] I have squashed any insignificant commits
- [x] This change has comments for package types, values, functions, and non-obvious lines of code
- [x] I have read the contribution requirements and fulfil them.

uehler commented on engine/Shopware/Components/CacheManager.php:
```diff
@@ -78,16 +80,23 @@ class CacheManager
+        ContainerAwareEventManager $events,
+        PathResolver $themePathResolver
+    ) {
         $this->container = $container;
```
You can pass the parameters like `shopware.httpCache.enabled` via constructor too and can get rid of the container injection

pascalheidmann-bedarf replied:
Yeah I know but wasn't sure as there already too many parameters 😅

mitelg replied:
I think, better many explicit parameters as one container parameter :wink:

mitelg replied:
you could try to inject `shopware.httpCache` afaik you will get an array, so you don't need `shopware.httpCache.enabled`, `shopware.httpCache.cache_dir`, etc

pascalheidmann-bedarf commented:
As suggested by @uehler and @mitelg I now inject all parameters (as arrays if possible) into CacheManager and removed the container completely.

mitelg commented on engine/Shopware/Components/CacheManager.php:
```diff
@@ -78,35 +86,78 @@ class CacheManager
      * Clear HTTP-Cache
      */
-    public function clearHttpCache()
+    public function clearHttpCache(): void
```
you should not change the method signature of public methods by adding typehints and return types. this will be a breaking change

mitelg commented on engine/Shopware/Components/CacheManager.php:
```diff
@@ -311,26 +352,21 @@ public function getConfigCacheInfo()
     /**
      * Returns template cache information
-     *
-     * @return array
```
do not delete return doc

mitelg commented on engine/Shopware/Components/CacheManager.php:
```diff
@@ -311,26 +352,21 @@ public function getConfigCacheInfo()
     /**
      * Returns template cache information
-     *
-     * @return array
```
do not delete return doc

mitelg commented on engine/Shopware/Components/CacheManager.php:
```diff
@@ -339,37 +375,28 @@ public function getThemeCacheInfo()
     /**
      * Returns cache information
-     *
-     * @return array
```
do not delete return doc

mitelg commented on engine/Shopware/Components/CacheManager.php:
```diff
@@ -339,37 +375,28 @@ public function getThemeCacheInfo()
     /**
      * Returns cache information
-     *
-     * @return array
```
do not delete return doc

mitelg commented on engine/Shopware/Components/CacheManager.php:
```diff
@@ -339,37 +375,28 @@ public function getThemeCacheInfo()
     /**
      * Returns cache information
-     *
-     * @return array
```
do not delete return doc

mitelg commented on engine/Shopware/Components/CacheManager.php:
```diff
@@ -391,15 +418,11 @@ public function getOpCacheCacheInfo()
      *
      * @param string $dir
-     *
-     * @return array
```
do not delete return doc

mitelg commented on engine/Shopware/Components/CacheManager.php:
```diff
@@ -452,13 +475,13 @@ public function getDirectoryInfo($dir)
      *
      * @param float $bytes
-     *
-     * @return string
```
do not delete return doc

mitelg commented on engine/Shopware/Components/CacheManager.php:
```diff
@@ -467,8 +490,6 @@ public function encodeSize($bytes)
      *
      * @param string[] $tags
-     *
-     * @return int
```
do not delete return doc

mitelg commented on engine/Shopware/Components/CacheManager.php:
```diff
@@ -485,8 +506,6 @@ public function clearByTags(array $tags)
      *
      * @param string $tag
-     *
-     * @return bool
```
do not delete return doc

soebbing commented on engine/Shopware/Components/CacheManager.php:
```diff
@@ -523,30 +564,30 @@ public function clearByTag($tag)
      * @param string $dir
      */
-    private function clearDirectory($dir)
+    private function clearDirectory($dir): void
```
You can add a string type here.

soebbing replied:
Please a string type here.

soebbing commented on engine/Shopware/Components/CacheManager.php:
```diff
@@ -239,16 +285,16 @@ public function clearOpCache()
-            );
+        if ($this->httpCache['enabled']) {
+            $info = $this->getDirectoryInfo($this->httpCache['cache_dir']);
         } else {
```
You can get rid of the `else` block here by simply initializing the variable `$info = [];` above the `if` condition. The same for the `backend` below.

pascalheidmann-bedarf commented:
@soebbing thx!
Because I was already changing more than signatures I added some deprecations for private methods and also removed indirect calls for `cache` service via `cacheManager` by replacing those with explicit service calls in the `Shopware\Components\Api\Resource\Cache` and cache backend controller.

Also I had to fix two return annotations in `\Shopware_Bootstrap` as those seem to be broken/detected now by PHPStan...

mitelg commented on engine/Library/Enlight/Components/Snippet/Manager.php:
```diff
@@ -75,6 +75,7 @@ public function __construct($options = null)
         if (isset($options['ignore_namespace'])) {
             $this->ignoreNamespace = (bool) $options['ignore_namespace'];
         }
+        parent::__construct();
```
are you sure, that this will not break anything?

pascalheidmann-bedarf replied:
`\Enlight_Components_Snippet_Manager` has only the abstract class `\Enlight_Class` as its ancestor - therefore only the branch `if (method_exists($this, 'init'))` might ever get executed.
So, yes, as long as no dev tries to extend `\Shopware_Components_Snippet_Manager` or `\Enlight_Components_Snippet_Manager` and adds an `init` method this won't have unwanted side-effects.

mitelg commented on engine/Shopware/Components/CacheManager.php:
```diff
@@ -78,20 +87,70 @@ class CacheManager
 
     /**
-     * @return \Zend_Cache_Core
+     * @deprecated in 5.7, will be removed in 5.8
```
you should describe an alternative for this call

mitelg commented on engine/Shopware/Components/CacheManager.php:
```diff
@@ -451,102 +548,49 @@ public function getDirectoryInfo($dir)
     /**
      * Format size method
      *
+     * @deprecated in 5.7, will be private in 5.8
```
alternative

pascalheidmann-bedarf replied:
I can't have any useful alternative but also see no reason for having this public. Right now the function is used just inside the class itself.

mitelg replied:
ah ok. then you should add this as info :wink: because it is not used in Shopware itself somewhere else, does not mean, some plugins don't use it :blush:

mitelg replied:
totally have overseen the private hint :smile: forget what i've said

mitelg commented on engine/Shopware/Components/CacheManager.php:
```diff
@@ -387,19 +429,74 @@ public function getOpCacheCacheInfo()
     /**
      * Returns cache information
      *
+     * @deprecated in 5.7, will be private in 5.8
```
alternative

pascalheidmann-bedarf replied:
This is a helper that is used only inside the class itself. I don't see any reason for an alternative, especially as this little formatter is just shy 4 lines of code. If a dev needs it somewhere they should better implement it themselves.
Only possible alternative would be moving it into a completely new helper service that wouldn't really fit anywhere.

soebbing commented on engine/Shopware/Components/Api/Resource/Cache.php:
```diff
@@ -74,13 +87,16 @@ public function getOne($id)
     /**
+     * @throws ApiException\PrivilegeException
+     * @throws NotFoundException
+     *
```
Please remove these Exception annotations.

The rule Shopware follows with regards to annotating Exceptions is: We only declare them when they are thrown in this method explicitly.
The reasoning for this is as follows: If I modify a function and want to declare all implicit Exceptions as well, it would be necessary to check all functions I am using in my modified function for possible exceptions (your IDE helps you there), but I would also need to change all method annotations for every method calling the modified one, since they would have to get the new exception annotation as well. And that is nearly impossible.

soebbing commented on engine/Shopware/Components/Api/Resource/Cache.php:
```diff
@@ -102,8 +118,10 @@ public function getList()
+     * @throws ApiException\PrivilegeException
+     * @throws NotFoundException
+     * @throws ParameterMissingException
+     * @throws \Zend_Cache_Exception
```
Please see the comment above

soebbing commented on engine/Shopware/Components/Snippet/Manager.php:
```diff
@@ -90,12 +90,12 @@ public function __construct(ModelManager $modelManager, array $pluginDirectories
         }
 
-        $this->adapter = new DbAdapter([
+        parent::__construct(new DbAdapter([
```
My Spidy-Developer senses are tingeling. Why did you change this? I don't see `$this->adapter` being defined anywhere, but it is used later on.

mitelg replied:
@soebbing see here. the parent has some "nice" possibilities of parameters in its constructor :smile: https://github.com/shopware/shopware/blob/5.6/engine/Library/Enlight/Components/Snippet/Manager.php#L63

soebbing replied:
Ok, I wish I wouldn't have seen that. 🤢 But I guess the change is ok then, technically!

pascalheidmann-bedarf replied:
@soebbing me neither. The main reason I stumbled upon this in the first was my IDE freaking out of missing parent constructor calls. I might add a comment to clarify this.

soebbing commented on engine/Shopware/Components/Api/Resource/Cache.php:
```diff
@@ -57,15 +68,17 @@ public function setContainer(Container $container = null)
-     * @throws \Shopware\Components\Api\Exception\NotFoundException
+     * @throws ApiException\PrivilegeException
+     * @throws NotFoundException               *@throws ApiException\PrivilegeException
+     * @throws ParameterMissingException
```
See comment below

pascalheidmann-bedarf commented:
While splitting all changes into their own commits I realised that the changes regarding SnippetManager aren't really part of the rest and therefore I moved them into #2195

shyim commented:
Wrong pull request number :man_facepalming:

shyim commented:
When I clear the cache I get following error message
"Fatal error: Uncaught PDOException: SQLSTATE[42S22]: Column not found: 1054 Unknown column 'routerlastupdate' in 'where clause' in /var/www/html/shopware/vendor/doctrine/dbal/lib/Doctrine/DBAL/Connection.php:912 Stack trace: #0 /var/www/html/shopware/vendor/doctrine/dbal/lib/Doctrine/DBAL/Connection.php(912): PDO->query('SELECT v.shop_i...') #1"

Please test your change before submitting your Pull Request :blush:

mitelg commented on engine/Shopware/Components/DependencyInjection/Bridge/Plugins.php:
```diff
@@ -24,24 +24,18 @@
-        \Enlight_Event_EventManager $eventManager,
-        \Shopware $application,
-        array $pluginDirectories
+        \Enlight_Loader $loader, \Enlight_Event_EventManager $eventManager, \Shopware $application, array $pluginDirectories, ConfigReader $configReader
```
please use multi lines here

shyim commented:
@pascalheidmann-bedarf Can i lookup? Maybe add next time new commits instead squashing, i can't see the difference

pascalheidmann-bedarf commented:
@shopwareBot I rebased to latest version and merged the conflicting changes in `\Shopware_Components_Translation`.
@shyim the mentioned problem happened cause I thought `Expr\Like` would escape... Added the explicit escaping. Next time I will definitely use more commits ;)

shyim commented:
Next time please test your PR before submitting it. It was just clearing cache :see_no_evil:

pascalheidmann-bedarf commented:
@shopwareBot pls use my private account @pascalheidmann for credits

shyim commented:
Thanks for your Contribution! :blue_heart:
